test/fgpu/test_engine.py

```diff
@@ -142,7 +138,36 @@ def output(self, output_arg: str) -> Output:
             return parse_narrowband(arg_value)
 
     @pytest.fixture
-    def engine_arglist(self, output_arg: str) -> list[str]:
+    def coherent_scale(self, output: Output) -> np.ndarray:
+        """Gain for tones.
+
+        Expected frequency-domain magnitude for a tone with time-domain
+        magnitude 1 when the eq gain is 1. The array is 1D, indexed by
+        channel.
+        """
+        pfb = generate_pfb_weights(output.spectra_samples // output.subsampling, output.taps, output.w_cutoff)
+        gain = np.repeat(np.sum(pfb), output.channels)
+        if isinstance(output, NarrowbandOutput):
+            ddc = generate_ddc_weights(output.ddc_taps, output.w_pass, output.w_stop, output.weight_pass)
+            response = np.fft.fftshift(scipy.signal.freqz(ddc, worN=output.spectra_samples, whole=True)[1])
+            # Discard higher frequencies
+            response = response[
+                (output.spectra_samples - output.channels) // 2 : (output.spectra_samples + output.channels) // 2
+            ]
+            gain *= np.abs(response)
+        # Factor of 2 is because the power is split between the positive and
+        # negative frequencies, and only the positive frequency is returned.
+        return gain / 2
```

Comment:
It's still true. The negative frequencies get filtered out by the low-pass filter.

Reply:
Maybe I'm not understanding something but in the wideband case the /2 makes sense because we drop the negative channels in the R2C PFB. So the loss of power there makes sense.

In the narrowband case though, you take the DDC filter into account earlier on. The "negative frequencies" are really only with reference to the centre frequency and we don't throw them away.

Unless there's something I'm missing. Maybe it would be better to clarify on Zoom?

Reply:
Let's say that your wideband bandwidth is 1000 MHz and the narrowband region is 400-500MHz, and consider a tone at 470MHz. It consists of half its power at +470MHz and half at -470MHz. First you multiply with a -450MHz mixer, which shifts those to +20MHz and -920MHz. Next you apply the low-pass filter (with passband of -50..+50MHz) which squashes the -920MHz tone.

If that still doesn't make sense, let's talk about it on Zoom.

Reply:
Right, I understand what you mean now. Makes more sense.